### PR TITLE
Graphene layer for serving the Runs Feed

### PR DESCRIPTION
**Pull request by jamiedemaria**
## Summary & Motivation
Introduces a way to query for the Runs Feed (backfills and runs that are not part of backfills).

Main entrypoint is the `GrapheneRunsFeedConnection` that returns a list of `GrapheneRunsFeedEntry`s,  a cursor, and whether more entries exist for future calls. 

`GrapheneRun` and `GraphenePartitionBackfill` both implement the `GrapheneRunsFeedEntry` interface, which should provide all of the attributes necessary to render the Runs Feed UI (@bengotow if you find during implementing the UI that something is missing, lmk) 

## How I Tested These Changes
new unit tests

### CODE REVIEW

**Comment by jamiedemaria** [comment]
* **#23375** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23375?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @jamiedemaria and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py
Comment (jamiedemaria) [inline python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py:427]: Do we need to worry about spicy window in the runs/bulk action table?

Screenshot 2024-08-06 at 10 01 07 AM

File: python_modules/dagster-graphql/dagster_graphql/schema/pipelines/pipeline.py
Comment (jamiedemaria) [inline python_modules/dagster-graphql/dagster_graphql/schema/pipelines/pipeline.py:627]: Naming thread! what do we want to call this?

- MegaRun
- BackfillOrRun
- LeadRun
- RunSet
- ???

File: python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py
Comment (jamiedemaria) [inline python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py:418]: cursor format: For querying runs, the cursor is the ID of the last run in the returned list. I started with something similar here where the cursor is the last run id and the last backfill id separated by a ";". The difficulty with that is that users would need to iterate through the returned list to find the last run/backfill. We could add a resolver on `GrapheneMegaRuns` that returns the cursor for them (i think that's reasonable), but i also want to think about cursor formats that might be easier to generate in the first place.

File: python_modules/dagster-graphql/dagster_graphql_tests/graphql/test_mega_run.py
Comment (jamiedemaria) [inline python_modules/dagster-graphql/dagster_graphql_tests/graphql/test_mega_run.py:90]: testing is not done yet. it would be really nice to be able to set the create_time for runs so that i can put the runs and backfills in the tables in a pre-specified order and then ensure the combined list shares that order, but i need to figure out how to do that

File: python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py
Comment (alangenfeld) [reply python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py:427]: technically the problem can occur but I wouldn't worry about it in this context

File: python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py
Comment (alangenfeld) [reply python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py:418]: heres what I did for similar problem recently https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_core/remote_representation/external.py#L85-L112

the cursor should be opaque to the client, it just gets echoed back to the server to fetch more

File: python_modules/dagster-graphql/dagster_graphql/schema/pipelines/pipeline.py
Comment (alangenfeld) [reply python_modules/dagster-graphql/dagster_graphql/schema/pipelines/pipeline.py:627]: instead of a new concrete type, it may be better to use an interface or a union and return the existing underlying types for run and backfill

https://graphql.com/learn/interfaces-and-unions/

File: python_modules/dagster-graphql/dagster_graphql/schema/runs.py
Comment (alangenfeld) [inline python_modules/dagster-graphql/dagster_graphql/schema/runs.py:228]: having an intermediate object here we should model this more as a "connection" than just a plural of results. Basically this just means returning cursor and maybe "has more" here. Also might inform a slightly different name. We are not consistent about this but if you grep for "connection" you can see some examples. 

https://graphql.org/learn/pagination/#complete-connection-model

File: python_modules/dagster-graphql/dagster_graphql_tests/graphql/test_mega_run.py
Comment (alangenfeld) [reply python_modules/dagster-graphql/dagster_graphql_tests/graphql/test_mega_run.py:90]: instead of asserting a fixed order could just assert that each result has a later creation time than the previous

**Review by alangenfeld — COMMENTED** [review]
especially if we move away from a new concrete type, we can focus the naming on the capability we are adding more ie: 

`mergedRuns`, `GrapheneMergedRunsConnection`, `GrapheneMergedRunInterface/Union` 

* merged runs
* grouped runs
* collapsed runs
* run feed (run feed entry/unit)

File: python_modules/dagster-graphql/dagster_graphql/schema/pipelines/pipeline.py
Comment (alangenfeld) [reply python_modules/dagster-graphql/dagster_graphql/schema/pipelines/pipeline.py:627]: Union is easiest from server but then client has to fork all rendering logic based on type , interface allows for a common set of fields to handle in one place. Talking with @bengotow and looking at what fields are common probably best way to navigate. Can always add new fields to the two concrete types to support the interface as well .

Comment (jamiedemaria) [reply python_modules/dagster-graphql/dagster_graphql/schema/pipelines/pipeline.py]: updated to use an interface

**Comment by jamiedemaria** [comment]
@alangenfeld i like the capability-focused naming option. I like `RunsFeed(Connection/Entry/...)`. Maybe it's a bit too narrow on the use case of displaying the runs feed, but i kind of like how clear and precise it is about its purpose

**Review by sryza — COMMENTED** [review]
> i like the capability-focused naming option. I like RunsFeed(Connection/Entry/...). Maybe it's a bit too narrow on the use case of displaying the runs feed, but i kind of like how clear and precise it is about its purpose

This makes a lot of sense to me.

FWIW I had been imagining that "mega run" would be the new name for what used to be called "backfill", and we'd use something like "outer run" to refer to what's called "mega run" in this PR, which is a run that's not a "sub-run".  Regular ol' runs don't seem particularly "mega" to me.

**Comment by jamiedemaria** [comment]
@alangenfeld @sryza I did a renaming pass to align everything on `RunsFeedConnection/Entry` and got everything switched over to use the connection model. Also modified the cursor a bit to account for another edge case. I think the sum of all the changes makes this ready for another proper review pass

File: python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py
Comment (alangenfeld) [inline python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py:447]: lets enforce limit as a required param instead

File: python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py
Comment (alangenfeld) [inline python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py:457]: nit: maybe a summary comment about having to loop to fill discarded backfill runs

File: python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py
Comment (alangenfeld) [inline python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py:479]: same as above - force limit set

File: python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py
Comment (alangenfeld) [inline python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py:493]: were looping here justto support created before filter?

File: python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py
Comment (alangenfeld) [inline python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py:566]: maybe add a method on the two python objects we can call instead ?

File: python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py
Comment (alangenfeld) [inline python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py:560]: nit: all_entries or smth to make it easier to follow that its not just DagsterRuns

File: python_modules/dagster-graphql/dagster_graphql/schema/backfill.py
Comment (alangenfeld) [inline python_modules/dagster-graphql/dagster_graphql/schema/backfill.py:317]: nit: maybe we should consolidate on to `timestamp` instead since its a `float` ? I feel like `time` is a better name for returning like a datetime object

File: python_modules/dagster-graphql/dagster_graphql/schema/backfill.py
Comment (alangenfeld) [inline python_modules/dagster-graphql/dagster_graphql/schema/backfill.py:360]: i dont think this enum provides any utility, we will have access to what the concrete types are

File: python_modules/dagster-graphql/dagster_graphql/schema/backfill.py
Comment (alangenfeld) [inline python_modules/dagster-graphql/dagster_graphql/schema/backfill.py:356]: hmm not sure about this part of the interface

**Review by alangenfeld — CHANGES_REQUESTED** [review]
nice this is close, worth a roundtrip i think tho

File: python_modules/dagster-graphql/dagster_graphql/schema/backfill.py
Comment (jamiedemaria) [reply python_modules/dagster-graphql/dagster_graphql/schema/backfill.py:356]: yeah i dont know what to do here either. i was thinking about naming it something more generic like `target` but then it seems like it should encompass `assetSelection` and `assetCheckSelection` too. maybe that would be ok?

File: python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py
Comment (jamiedemaria) [reply python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py:493]: yep, it's not ideal. there isn't support for filtering backfills on anything other than status. I want to add more filter support for backfills, but didn't want to expand this PR. Planning to stack a branch that adds more filtering capabilities

File: python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py
Comment (jamiedemaria) [reply python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py:566]: the two objects being `GrapheneRun` and `GraphenePartitionBackfill` or `RunRecord` and `PartitionBackfill`?

File: python_modules/dagster-graphql/dagster_graphql/schema/backfill.py
Comment (jamiedemaria) [reply python_modules/dagster-graphql/dagster_graphql/schema/backfill.py:317]: i like that

File: python_modules/dagster-graphql/dagster_graphql/schema/backfill.py
Comment (alangenfeld) [reply python_modules/dagster-graphql/dagster_graphql/schema/backfill.py:356]: in the end the question is how would this field be used on the client - so best path is to get @bengotow 's input either in this initial PR or maybe defer it to a second iteration

File: python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py
Comment (alangenfeld) [reply python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py:566]: ah right these are the graphene objects, yea i think it would still feel a bit cleaner to add a `@property` `creation_time` to call here and then have their `resolve_` methods call in to as well but thats more of a nit pick so leave to your call

Comment (jamiedemaria) [reply python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py:566]: other option would be to hold off on converting to the `Graphene...` objects until the very end and then iterate through the final list of objects to return and do the conversion

Comment (jamiedemaria) [reply python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py:566]: and add a shared prop to both  `RunRecord` and `PartitionBackfill`

File: python_modules/dagster-graphql/dagster_graphql/schema/backfill.py
Comment (jamiedemaria) [reply python_modules/dagster-graphql/dagster_graphql/schema/backfill.py:317]: my thinking with keeping createTime, startTime, endTime was that in the future where everything can be a single run, we won't have these duplicate fields on GrapheneRun

File: python_modules/dagster-graphql/dagster_graphql/schema/backfill.py
Comment (alangenfeld) [inline python_modules/dagster-graphql/dagster_graphql/schema/backfill.py:128]: probably want to edit comment if this is going to land and not be just PR time placeholder

**Review by alangenfeld — APPROVED** [review]

File: js_modules/dagster-ui/packages/ui-core/src/graphql/schema.graphql
Comment (bengotow) [inline js_modules/dagster-ui/packages/ui-core/src/graphql/schema.graphql:2774]: This schema looks great but giving backfills a `runId` feels a bit icky -- I wonder if we could call the common interface's id and status just `id` and `status`, or potentially make new `entryId` and `entryStatus`?

Alternatively it might be nice to put a Grafene description on these fields to clarify that they're just meant to implement this common interface

**Review by bengotow — APPROVED** [review]
Will defer to others on the Python bits but the new schema looks good to me! Just left one inline comment

File: js_modules/dagster-ui/packages/ui-core/src/graphql/schema.graphql
Comment (jamiedemaria) [reply js_modules/dagster-ui/packages/ui-core/src/graphql/schema.graphql:2774]: id and status are already taken (by the backfill id and bulk action status respectively). My hesitation with adding `entryId` and `entryStatus` is that we'd have to add those to `GrapheneRun` as well. I was hoping to keep `GrapheneRun` as unchanged as possible so that in the future when we've eliminated the need for backfills entirely, the `GrapheneRun` doesn't have attributes that are there because of an interface that is no longer necessary

Comment (jamiedemaria) [reply js_modules/dagster-ui/packages/ui-core/src/graphql/schema.graphql:2774]: I'll definitely add descriptions though. I'm also open to pushback on not doing `entryId/Status`. If it adds significant confusion/code smell it might not be worth it

Comment (jamiedemaria) [reply js_modules/dagster-ui/packages/ui-core/src/graphql/schema.graphql:2774]: @alangenfeld  were you thinking we'd switch GraphenRun.id of GraphenePartitionBackfill.id to either string or ID so that the types match?

Comment (jamiedemaria) [reply js_modules/dagster-ui/packages/ui-core/src/graphql/schema.graphql:2774]: > the status one i think is useful since it flattens to just one enum

can you elaborate on this a bit?

Comment (alangenfeld) [reply js_modules/dagster-ui/packages/ui-core/src/graphql/schema.graphql:2774]: > were you thinking we'd switch GraphenRun.id of GraphenePartitionBackfill.id to either string or ID so that the types match?
 
I assumed they were the same, i think its fine to change backfill.id to `ID` to align them, pretty sure that a safe change since ID is String

> can you elaborate on this a bit?

as opposed to the id change above, we cant change the return type of either existing status field so we need to add a new field to return one enum type on both objects. That assumes that we want that which @bengotow can speak to

**Pull request merged**
